### PR TITLE
Store the client version in the database.

### PR DESCRIPTION
**Pull request by Pentarctagon**
This will now store the version of the client playing a side in `game_player_info`.

### CODE REVIEW

**Comment by Pentarctagon** [comment]
@Vultraz and @soliton-:
While there are a few issues that remain to be worked out on the second commit, with both of them combined, they will mean we would have solid data to look at as far as what versions people are using on the MP server and where they got them.

Thoughts?

**Comment by soliton-** [comment]
We can already check the logs for what versions are used but putting it in the DB is good, too.

Recording how the client was distributed sounds interesting.

**Comment by Pentarctagon** [comment]
The code at least is fine now, I believe.

**Comment by Pentarctagon** [comment]
Also, since this will require Wesnoth's packagers to set explicitly this flag for it to be useful:

@loonycyborg @hrubymar10 @singalen @rhonda (I'm not sure of alep's account name for Android) - 

The second commit of this PR adds the `CLIENT_SOURCE` define, which is intended to be used so we can know where people who play on the official multiplayer server downloaded Wesnoth from.  This can be set for cmake via `-DCLIENT_TYPE=<N>`, scons via `build_type=<N>`, and I have to assume xcode has a way to allow adding this as well.  The list of valid values can be found [here](https://github.com/wesnoth/wesnoth/blob/dca8badf9b1de836163a5a64824475622809c5aa/src/game_version.cpp#L37-L59).

Can you all set this define to the appropriate value going forward when you create packages for Wesnoth?

**Comment by hrubymar10** [comment]
@Pentarctagon I don't like the idea of creating 3 different builds. I am creating one build right now which is distributed to the 3 different sources. Can you support for example `dist` file inside root of data?

Inside this dist file will be your enum number so I won't have to create 3 different builds but I will just manually change this number for each package.

Also you can make `CLIENT_SOURCE == 1` the default one thanks to this.

**Comment by Pentarctagon** [comment]
Sure, it can work that way instead.

**Comment by jostephd** [comment]
> Inside this dist file will be your enum number

Might as well make it a string rather than an enum, at this point.

**Comment by Pentarctagon** [comment]
It's not really an enum either way, but I'd rather not just send over and insert whatever the contents of the `dist` file is, given it's easily editable by anyone.

**Comment by jostephd** [comment]
> It's not really an enum either way, but I'd rather not just send over and insert whatever the contents of the `dist` file is, given it's easily editable by anyone.

Then make the content a string, and validate it against a whitelist of nine or ten hardcoded strings before doing anything else with it.

**Comment by Pentarctagon** [comment]
@hrubymar10 Updated.  The list of values is now [here](https://github.com/wesnoth/wesnoth/blob/1692a0c6fb7a6cb2f5e849c6528b0c77491d8f09/src/game_initialization/multiplayer.cpp#L159-L177).

**Comment by hrubymar10** [comment]
Very well,
Thank you very much 😊

This will make it much easier for me.

BTW is this going to be merged before 1.15.2?

**Comment by Pentarctagon** [comment]
Ideally yes, depending on when other people respond, but it's not something I consider urgent either.  Mostly I'd like to get this in before a 1.14.10 release.

**Comment by jostephd** [comment]
May I reiterate my recommendation to make the content of the file a string?

**Comment by Pentarctagon** [comment]
I have changed it to that locally after loonycyborg and singalen said the same, I just haven't pushed it here since I haven't re-tested it yet.

**Comment by jostephd** [comment]
:+1:. For future reference, even if it had been a -Dfoo=bar, I would still have recommended using strings. There would have been multiple ways to implement that.

**Comment by Pentarctagon** [comment]
Updated to use strings.

File: src/server/server.cpp [hunk 1]
```diff
@@ -590,14 +590,17 @@ void server::read_version(socket_ptr socket, std::shared_ptr<simple_wml::documen
 		const simple_wml::string_span& version_str_span = (*version)["version"];
 		const std::string version_str(version_str_span.begin(), version_str_span.end());
 
+		const simple_wml::string_span& source_str_span = (*version)["client_source"]; // TODO: handle this attribute not being present on previous versions of the client
```
Comment (soliton-) [inline src/server/server.cpp]: This will just be empty for old clients which seems fine, no?

File: src/server/server.cpp [hunk 2]
```diff
@@ -1620,7 +1625,19 @@ void server::handle_player_in_game(socket_ptr socket, std::shared_ptr<simple_wml
 			const simple_wml::node::child_list& sides = g.get_sides_list();
 			for(unsigned side_index = 0; side_index < sides.size(); ++side_index) {
 				const simple_wml::node& side = *sides[side_index];
-				user_handler_->db_insert_game_player_info(uuid_, g.id(), side["player_id"].to_string(), side["side"].to_int(), side["is_host"].to_bool(), side["faction"].to_string());
+				const auto player = player_connections_.get<name_t>().find(side["player_id"].to_string());
```
Comment (soliton-) [inline src/server/server.cpp]: I wonder if the find call is necessary or this should iterate over something else.

It‘s just on game start though so shouldn’t be a big deal.

File: src/server/server.cpp (hunk 1 above)
Comment (Pentarctagon) [reply src/server/server.cpp]: Ah, yeah, I had tested that but apparently forgot to remove to TODO comment I left myself.

File: src/server/server.cpp (hunk 2 above)
Comment (Pentarctagon) [reply src/server/server.cpp]: I'm not sure where else I'd get the version/source info from, either.

Comment (soliton-) [reply src/server/server.cpp]: Getting the info from the player is fine, it’s just a bit weird to have to search. Could perhaps iterate over the players instead but maybe you‘d miss empty sides then or something. Haven’t looked more closely.

Comment (Pentarctagon) [reply src/server/server.cpp]: Searching a multi-index by one of its indexed values should be pretty fast though, right?

Comment (soliton-) [reply src/server/server.cpp]: Yeah, in fact it should not be a search at all then. Maybe find works that way here. Still not very familiar with it.

**Pull request merged**